Pass response object to genReqId()

bmodotdev opened the pull request:
Passing the response body object to genReqId() allows us to more easily
share the request ID throughout our code, e.g. setting an “X-Request-Id”
header via “response.header()” or storing it in “response.locals” for
rendering.

bmodotdev commented:
This feels a like silly commit given its size, but I believe providing the response object makes a lot of sense. I can see `req` makes sense when the client is querying the server with an already existing `X-Request-Id` header, so you simply take it from the `req` object.

However, in the case that a request id does not exist in the `req` object, if you generate one here in your logger, you must store it in the `req` object for later reference, then later pull it out for setting the header.

To workaround this, I just created a UUID middleware that generates the id, sets the header, then stores it in `req` for exposing to the logger via `genReqId()`.

Ideally, I would not need a middleware, `genReqId()` would simply pass the `req` and `res` objects and I simply generate it there, set it in headers, and return it to the logger, all in one place.

Anyway, just a small improvement. Thanks for the awesome library!

mcollina reviewed (changes requested):
Thanks for opening a PR! Can you please add a unit test?

bmodotdev commented:
> Thanks for opening a PR! Can you please add a unit test?

I am having a hard time thinking of how a unit test for passing express response object might function to `genReqId()`. Passing the response object into `genReqId()` is moreso for the benefit of other code, i.e. setting a header or setting `res.locals`. I have updated the README though, and updated the unit tests to pass `res`.

If that is not sufficient, should I just use `res` to set a header or set a `res.locals.id`, then check it? Sorry, still understanding the codebase.

mcollina commented on README.md:
```diff
@@ -122,13 +122,21 @@ $ node example.js | pino-pretty
 
 const http = require('http')
 const server = http.createServer(handle)
+const { v4: uuidv4 } = require('uuid')
```
Could you recommend https://nodejs.org/api/crypto.html#cryptorandomuuidoptions instead? It's _faster_ and does not require external deps.

bmodotdev replied:
absolutely, thanks for the suggestion, this is done

mcollina commented on test/test.js:
```diff
@@ -290,7 +290,7 @@ test('uses a custom genReqId function', function (t) {
   const dest = split(JSON.parse)
   let idToTest
-  function genReqId (req) {
+  function genReqId (req, res) {
```
Please add an assertion to verify that `res` is set.

bmodotdev replied:
This is done

mcollina commented on test/test.js:
```diff
@@ -536,7 +536,7 @@ test('support a custom instance with custom genReqId function', function (t) {
 
   let idToTest
-  function genReqId (req) {
+  function genReqId (req, res) {
```
Please add an assertion to verify that `res` is set.

bmodotdev replied:
This is done

mcollina reviewed (approved):
lgtm

bmodotdev commented:
Awesome, thanks for the guidance. That simple assertion check makes much more since now, I was trying to over complicate it :)

MrTalon63 commented:
GenReqId interface does not have `res` specified resulting in typescript error
